near-sdk-macros/src/core_impl/code_generator/attr_sig_info.rs

```diff
@@ -191,10 +192,40 @@ impl AttrSigInfo {
                         }
                     }
                     BindgenArgType::CallbackResultArg => {
-                        let deserialize = deserialize_data(serializer_ty);
+                        let ok_type = extract_ok_type(ty).unwrap_or_else(||
+                            panic!("Function parameters marked with #[callback_result] should have \
+                                type Result<T, PromiseError>")
+                        );
+
+                        let deserialization_branch = match ok_type {
+                            // The unit type in this context is a bit special because functions
+                            // without an explicit return type do not serialize their response.
+                            // But when someone tries to refer to their callback result with
+                            // `#[callback_result]` they specify the callback type as
+                            // `Result<(), PromiseError` which cannot be correctly deserialized from
+                            // an empty byte array.
+                            //
+                            // So instead of going through serde, we do a manual check
+                            // that the byte array is either empty or contains the string "null".
+                            syn::Type::Tuple(type_tuple) if type_tuple.elems.is_empty() => {
+                                quote! {
+                                    near_sdk::PromiseResult::Successful(data) if data.is_empty() || data == vec!(110, 117, 108, 108) =>
```

Comment:
```suggestion
                                    near_sdk::PromiseResult::Successful(data) if data.is_empty() || &data == b"null" =>
```

`vec!` will allocate, and we don't need this to check equality.

Also, `null` will be deserialized into `()`, so this code path isn't really necessary. Also, I'm pretty sure this will fail when using another serialization protocol, try it with `borsh`

Reply:
All good points. I made so that the flow goes through the regular deserialization process if the array is not empty, this should resolve all the weirdness you have pointed out.